master/docs/developer/rest.rst

```diff
@@ -0,0 +1,236 @@
+For example:
+
+* ``http://build.example.org/api/v2/buildrequest?order=builderid&order=buildrequestid``
+
+Pagination
+..........
+
+Collection responses may be paginated with the ``offset`` and ``limit`` query parameters.
+The offset is the 0-based index of the first result to included, after filtering and sorting.
+The limit is the maximum number of results to return.
+Some resource types may impose a maximum on the limit parameter; be sure to check the resulting links to determine whether further data is available.
+For example:
+
+* ``http://build.example.org/api/v2/buildrequest?order=builderid&limit=10``
+* ``http://build.example.org/api/v2/buildrequest?order=builderid&offset=20&limit=10``
+
+Controlling
+~~~~~~~~~~~
+
+Data API control operations are handled by POST requests using a simplified form of `JSONRPC 2.0 <http://www.jsonrpc.org/specification>`_.
+The JSONRPC "method" is mapped to the data API "action", and the parameters are passed to that application.
+
+The following parts of the protocol are not supported:
+
+* positional parameters
+* batch requests
+
+Requests are sent as an HTTP POST, containing the request JSON in the body.
+The content-type header must be ``application/json``.
+
+A simple example:
+
+.. code-block:: none
+
+    POST http://build.example.org/api/v2/scheduler/4
+    --> {"jsonrpc": "2.0", "method": "force", "params": {"revision": "abcd", "branch": "dev"}, "id": 843}
+    <-- {"jsonrpc": "2.0", "result": {"buildsetid": 44}, "id": 843}
+
+.. _Raml-Spec:
+
+Raml Specs
+~~~~~~~~~~
+
+The Data API is documented in `RAML 1.0 format <https://github.com/raml-org/raml-spec/blob/raml-10/versions/raml-10/raml-10.md>`_.
+RAML describes and documents all our data, rest, and javascript APIs in a format that can be easily manipulated by human and machines.
+
+.. toctree::
+    :maxdepth: 2
+
+.. jinja:: data_api
+
+    {% for name, type in raml.types.items()|sort %}
+    ..
+       sphinx wants to have at least same number of underline chars than actual tile
+       but has the title is generated, this is a bit more complicated.
+       So we generate hundred of them
+
+    {{type.get("displayName", name)}}
+    {{"."*100}}
+
+    .. bb:rtype:: {{name}}
+
+        {% if 'properties' in type -%}
+        {% for key, value in type.properties.items() -%}
+        :attr {{value.type}} {{key}}: {{raml.reindent(value.description, 4*2)}}
+        {% endfor %}
+    {% if 'example' in type -%}
+
+    ``example``
+
+        .. code-block:: javascript
+
+            {{raml.format_json(type.example, indent=4*2)}}
+
+    {% endif %}
+    {% if 'examples' in type -%}
+    ``examples``
+
+    {% for example in type.examples -%}
+
+        .. code-block:: javascript
+
+            {{raml.format_json(example, indent=4*2)}}
+
+    {% endfor %}
+    {% endif %}
+
+    {{type.description}}
+    {% endif %}
+    {% if name in raml.endpoints_by_type -%}
+    Endpoints
```

Comment:
look like a endfor is missing here. the endfor line 236 is about the type (``{% for name, type in raml.types.items()|sort %}``)